### PR TITLE
Feature/gradle/enhance update java compatibility

### LINKED ISSUE #3235: Gradle recipe which sets Java sourceCompatibility and targetCompatibility
Similar to what we have for maven. For inclusion in the java 8 -> 11 and java 11 -> 17 migrations

### PR DESCRIPTION
**Pull request by shanman190**
Fixes #3235.

### CODE REVIEW

**Comment by shanman190** [comment]
The plan is for this to be folded into the `UpdateJavaVersion` recipe found in `rewrite-migrate-java`, so that it's able to handle even more Gradle options for Java migrations.

**Comment by rpau** [comment]
Thanks for the contribution. Now it is a little bit late for us to review the PR. I am adding my colleagues as reviewers to take a look in the next days.

File: rewrite-gradle/src/main/java/org/openrewrite/gradle/UpdateJavaCompatibility.java [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+    protected TreeVisitor<?, ExecutionContext> getSingleSourceApplicableTest() {
+        return new IsBuildGradle<>();
+    }
+
+    @Override
+    public TreeVisitor<?, ExecutionContext> getVisitor() {
+        return new GroovyVisitor<ExecutionContext>() {
+            final MethodMatcher sourceCompatibilityDsl = new MethodMatcher("RewriteGradleProject setSourceCompatibility(..)");
+            final MethodMatcher targetCompatibilityDsl = new MethodMatcher("RewriteGradleProject setTargetCompatibility(..)");
+            final MethodMatcher javaLanguageVersionMatcher = new MethodMatcher("org.gradle.jvm.toolchain.JavaLanguageVersion of(int)");
```
Comment (timtebeek) [inline rewrite-gradle/src/main/java/org/openrewrite/gradle/UpdateJavaCompatibility.java:83]: Nice addition here as well, thanks!

**Review by timtebeek — APPROVED** [review]
Looks good overall, as compared to the previous implementation. I think with 8.0 coming up it's fine to have this as a breaking change, especially as I'm [not seeing it referenced elsewhere](https://github.com/search?q=org%3Aopenrewrite%20ChangeJavaCompatibility&type=code).

My only change suggestion would have been to use a parameterized test to cover more of the argument & output combinations, but I'm not going to ask that you change that now. Appreciate the help!

File: rewrite-gradle/src/main/java/org/openrewrite/gradle/UpdateJavaCompatibility.java [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+    CompatibilityType compatibilityType;
+
+    @Option(displayName = "Declaration Style",
+            description = "The desired style to write the new version as when being written to the `sourceCompatibility` " +
+                    "or `targetCompatibility` variables. Default, match current source style. " +
+                    "(ex. Enum: `JavaVersion.VERSION_11`, Number: 11, or String: \"11\")",
+            valid = {"Enum", "Number", "String"},
+            required = false)
+    @Nullable
+    DeclarationStyle declarationStyle;
```
[the remark this answers is not in the document]
Comment (shanman190) [reply rewrite-gradle/src/main/java/org/openrewrite/gradle/UpdateJavaCompatibility.java:56]: Yeah, it was an early journey recipe and as you can imagine, there are new features and inspiration to work from. 😄

**Comment by shanman190** [comment]
@timtebeek, I very much appreciate the feedback. I'm going to go ahead and add at least a few of the combination tests as well, so there'll be an update here pretty shortly. 😄

**Pull request merged**